storage/accumulator/src/lib.rs

```diff
@@ -205,27 +214,33 @@ where
                 pos = pos.parent();
                 to_freeze.push((pos, hash));
             }
+            // The node remaining now must be a left child, possibly a complete binary tree.
             left_siblings.push((pos, hash));
         }
 
-        let (mut pos, mut hash) = left_siblings.pop().expect("Must have at least one node");
-        for _ in pos.level()..root_level as u32 {
-            hash = if pos.is_left_child() {
-                Self::hash_internal_node(hash, *ACCUMULATOR_PLACEHOLDER_HASH)
-            } else {
-                let sibling = pos.sibling();
-                match left_siblings.pop() {
-                    Some((x, left_hash)) => {
-                        assert_eq!(x, sibling);
-                        Self::hash_internal_node(left_hash, hash)
-                    }
-                    None => Self::hash_internal_node(self.reader.get(sibling)?, hash),
-                }
-            };
-            pos = pos.parent();
-        }
-
-        Ok((hash, to_freeze))
+        // Now reconstruct the final root hash by walking up to root level and adding
+        // placeholder hash nodes as needed on the right, and left siblings that have either
+        // been newly created or read from storage.
+        let (mut pos, hash) = left_siblings.pop().expect("Must have at least one node");
+        let res: Result<(HashValue, Vec<Node>)> = (pos.level()..root_level as u32)
+            .into_iter()
+            .try_fold(hash, |h, _| {
+                let h_new = if pos.is_left_child() {
+                    Self::hash_internal_node(h, *ACCUMULATOR_PLACEHOLDER_HASH)
+                } else {
+                    Self::hash_internal_node(
+                        left_siblings
+                            .pop()
+                            .map_or_else(|| self.reader.get(pos.sibling()), |(_, x)| Ok(x))?,
+                        h,
+                    )
+                };
+                pos = pos.parent();
+                Ok(h_new)
+            })
+            .map(|h| (h, to_freeze));
+        assert!(left_siblings.is_empty());
```

Comment:
This assert is bogus in the error case and the fold implementation doesn't appear to be any cleaner; I think I will revert this and just leave the comments.